drivers/i2s/i2s_mcux.c

```diff
@@ -0,0 +1,1209 @@
+	struct stream *strm = &dev_data->rx;
+	void *buffer;
+	int ret = 0;
+
+	LOG_DBG("i2s_mcux_read");
+	if (strm->state == I2S_STATE_NOT_READY) {
+		LOG_ERR("invalid state %d", strm->state);
+		return -EIO;
+	}
+
+	ret = k_msgq_get(&strm->out_queue, &buffer,
+			 SYS_TIMEOUT_MS(strm->cfg.timeout));
+	if (ret != 0) {
+		LOG_DBG("need retry");
+		return -EAGAIN;
+	}
+
+	switch (strm->state) {
+	case I2S_STATE_STOPPING:
+		i2s_rx_stream_disable(dev);
+		strm->state = I2S_STATE_READY;
+		break;
+	case I2S_STATE_RUNNING:
+		SAI_RxEnableDMA(DEV_CFG(dev)->base,
+				kSAI_FIFORequestDMAEnable, true);
+		break;
+	default:
+	case I2S_STATE_READY:
+		break;
+	}
+	*mem_block = buffer;
+	*size = strm->cfg.block_size;
+	return 0;
+}
+
+static int i2s_mcux_write(const struct device *dev, void *mem_block,
+			  size_t size)
+{
+	struct i2s_dev_data *dev_data = DEV_DATA(dev);
+	struct stream *strm = &dev_data->tx;
+	int ret;
+
+	LOG_DBG("i2s_mcux_write");
+	if (strm->state != I2S_STATE_RUNNING &&
+	    strm->state != I2S_STATE_READY) {
+		LOG_ERR("invalid state (%d)", strm->state);
+		return -EIO;
+	}
+
+	ret = k_msgq_put(&strm->in_queue, &mem_block,
+			 SYS_TIMEOUT_MS(strm->cfg.timeout));
+	if (ret) {
+		LOG_ERR("k_msgq_put failed %d", ret);
+		return ret;
+	}
+
+	if (strm->state == I2S_STATE_RUNNING && strm->stream_starving) {
+		/* in running state if the stream is off turn it on*/
+		ret = i2s_tx_stream_start(dev);
+		strm->stream_starving = false;
+	}
+
+	return ret;
+}
+
+static void sai_driver_irq(const struct device *dev)
+{
+	if ((DEV_BASE(dev)->TCSR & I2S_TCSR_FEF_MASK) == I2S_TCSR_FEF_MASK) {
+		/* Clear FIFO error flag to continue transfer */
+		SAI_TxClearStatusFlags(DEV_BASE(dev), I2S_TCSR_FEF_MASK);
+
+		/* Reset FIFO for safety */
+		SAI_TxSoftwareReset(DEV_BASE(dev), kSAI_ResetTypeFIFO);
+
+		LOG_DBG("sai tx error occurred");
+	}
+
+	if ((DEV_BASE(dev)->RCSR & I2S_RCSR_FEF_MASK) == I2S_RCSR_FEF_MASK) {
+		/* Clear FIFO error flag to continue transfer */
+		SAI_RxClearStatusFlags(DEV_BASE(dev), I2S_RCSR_FEF_MASK);
+
+		/* Reset FIFO for safety */
+		SAI_RxSoftwareReset(DEV_BASE(dev), kSAI_ResetTypeFIFO);
+
+		LOG_DBG("sai rx error occurred");
+	}
+}
+
+/* clear IRQ sources atm */
+static void i2s_mcux_isr(void *arg)
+{
+	struct device *dev = (struct device *)arg;
+
+	if ((DEV_BASE(dev)->RCSR & I2S_TCSR_FEF_MASK) == I2S_TCSR_FEF_MASK) {
+		sai_driver_irq(dev);
+	}
+
+	if ((DEV_BASE(dev)->TCSR & I2S_RCSR_FEF_MASK) == I2S_RCSR_FEF_MASK) {
+		sai_driver_irq(dev);
+	}
+	/*
+	 * Add for ARM errata 838869, affects Cortex-M4,
+	 * Cortex-M4F Store immediate overlapping exception return operation
+	 *  might vector to incorrect interrupt
+	 */
+#if defined __CORTEX_M && (__CORTEX_M == 4U)
+	__DSB();
+#endif
+}
+
+static void audio_clock_settings(const struct device *dev)
+{
+	uint32_t lp, pd, num, den, src, clK_src, pre_div, src_div;
+	clock_audio_pll_config_t audioPllConfig;
+
+	if (DEV_CFG(dev)->bus_id == 0) {
+#ifdef CONFIG_I2S_0
```

Comment:
This can be replaced with Device Tree `DT_INST `Defines.

Reply:
let me correct.